hagelslag/processing/STObject.py

```diff
@@ -101,6 +101,92 @@ def center_of_mass(self, time):
+                                                                                 self.j[diff].ravel()[valid])
+            else:
+                com_i = np.mean(self.i[diff])
+                com_j = np.mean(self.j[diff])
+            com_i = int(np.round(com_i))
+            com_j = int(np.round(com_j))
+        else:
+            com_i = None
+            com_j = None
+        return com_i, com_j
+
+    def extract_patch(self, patch_radius, full_x, full_y, full_i, full_j):
+        """
+        Extract patch of uniform radius from existing STObject. This is intended for extracting patches from
+        STObjects that are built around the bounding box of the original object. Areas outside the object are
+        padded with zeros.
+
+        Args:
+            patch_radius (int): radius of patch in pixels
+            full_x: full x grid encompassing the original field from which the objects have been extracted.
+            full_y: full y grid
+            full_i: full i (row) grid
+            full_j: full j (row) grid
+
+        Returns:
+            new STObject containing the a patched slice of the original STObject.
+        """
+        com_ijs = [self.center_of_mass_ij(time) for time in self.times]
+        patch_grid = []
+        patch_mask = []
+        patch_x = []
+        patch_y = []
+        patch_i = []
+        patch_j = []
+        for t, time in enumerate(self.times):
+            obj_slice_buff = (slice(com_ijs[t][0] - patch_radius, com_ijs[t][0] + patch_radius),
+                              slice(com_ijs[t][1] - patch_radius, com_ijs[t][1] + patch_radius))
+            obj_slice_local = (slice(com_ijs[t][0] - self.i[t].min() - patch_radius,
+                                     com_ijs[t][0] - self.i[t].min() + patch_radius),
+                               slice(com_ijs[t][1] - self.j[t].min() - patch_radius,
+                                     com_ijs[t][1] - self.j[t].min() + patch_radius))
+            patch_i.append(full_i[obj_slice_buff])
+            patch_j.append(full_j[obj_slice_buff])
+            patch_x.append(full_x[obj_slice_buff])
+            patch_y.append(full_y[obj_slice_buff])
+            pad_i_l = abs(obj_slice_local[0].start) if obj_slice_local[0].start < 0 else 0
+            pad_i_u = obj_slice_local[0].stop - self.timesteps[t].shape[0] \
+                if obj_slice_local[0].stop - self.timesteps[t].shape[0] > 0 else 0
+            pad_j_l = abs(obj_slice_local[1].start) if obj_slice_local[1].start < 0 else 0
+            pad_j_u = obj_slice_local[1].stop - self.timesteps[t].shape[1] \
+                if obj_slice_local[1].stop - self.timesteps[t].shape[1] > 0 else 0
+            if obj_slice_local[0].start < 0:
+                obj_slice_local[0].start = 0
+                obj_slice_local[0].stop += pad_i_l
+            if obj_slice_local[1].start < 0:
+                obj_slice_local[1].start = 0
+                obj_slice_local[1].stop += pad_j_l
```

Comment:
Slice object attributes appear to be read only. So it looks like you might need to fully redefine the slice or define it after the conditional.